relayer/relays/execution/message/message.go

```diff
@@ -0,0 +1,320 @@
+	}
+
+	return resultMessages
+}
+
+func filterMessagesByLastNonces(messages []*chain.EthereumOutboundMessage, addressNonceMap map[common.Address]uint64) []*chain.EthereumOutboundMessage {
+	resultMessages := []*chain.EthereumOutboundMessage{}
+
+	for _, basicMessage := range messages {
+		if basicMessage.Nonce <= addressNonceMap[basicMessage.Origin] {
+			continue
+		}
+
+		resultMessages = append(resultMessages, basicMessage)
+	}
+
+	return resultMessages
+}
+
+func (m *Message) syncUnprocessedBasicMessages(ctx context.Context) (uint64, error) {
+	lastVerifiedBlockNumber, err := m.writer.GetLastBasicChannelBlockNumber()
+	if err != nil {
+		return 0, fmt.Errorf("fetch last basic channel message block number: %w", err)
+	}
+
+	addressNonceMap, err := m.writer.GetLastBasicChannelNonceByAddresses(m.addresses)
+	if err != nil {
+		return 0, fmt.Errorf("fetch last basic channel message nonce: %w", err)
+	}
+
+	addressNonzeroNonceMap := make(map[common.Address]uint64, len(addressNonceMap))
+	for address, nonce := range addressNonceMap {
+		log.WithFields(log.Fields{
+			"block_number": lastVerifiedBlockNumber,
+			"address":      address,
+			"nonce":        nonce,
+		}).Info("last basic channel")
+
+		if nonce != 0 {
+			addressNonzeroNonceMap[address] = nonce
+		}
+	}
+
+	if len(addressNonzeroNonceMap) == 0 {
+		return 0, nil
+	}
+
+	log.WithFields(log.Fields{
+		"block_number": lastVerifiedBlockNumber,
+		"nonces":       addressNonzeroNonceMap,
+	}).Info("checking last synced basic channel messages on startup")
+	basicPayload, err := m.listener.ProcessBasicEvents(ctx, lastVerifiedBlockNumber, lastVerifiedBlockNumber, addressNonzeroNonceMap)
+	if err != nil {
+		return 0, err
+	}
+
+	basicPayload.Messages = filterMessagesByLastNonces(basicPayload.Messages, addressNonzeroNonceMap)
+
+	err = m.writeBasicMessages(ctx, basicPayload)
+	if err != nil {
+		return 0, err
+	}
+
+	return lastVerifiedBlockNumber, err
+}
+
+func (m *Message) syncUnprocessedIncentivizedMessages(ctx context.Context) (uint64, error) {
+	lastVerifiedBlockNumber, err := m.writer.GetLastIncentivizedChannelBlockNumber()
+	if err != nil {
+		return 0, fmt.Errorf("fetch last incentivized channel message block number: %w", err)
+	}
+
+	nonce, err := m.writer.GetLastIncentivizedChannelNonce()
+	if err != nil {
+		return 0, fmt.Errorf("fetch last incentivized channel message nonce: %w", err)
+	}
+
+	log.WithFields(log.Fields{
+		"block_number": lastVerifiedBlockNumber,
+		"nonce":        nonce,
+	}).Info("checking last synced incentivized channel messages on startup")
+
+	// If the last nonce is set, there could be messages that have not been processed in the same block.
+	// Messages that have already been verified will not be reprocessed because they will be filtered out
+	// in filterMessagesByLastNonce.
+	// Messages after the lastVerifiedBlockNumber will be processed normally in the go routine below.
+	if nonce == 0 {
+		return 0, nil
```

Comment:
When the nonce is zero, there aren't any messages to relay up to and including `lastVerifiedBlockNumber`, so we should be able to return that here.

```suggestion
	if nonce == 0 {
		return lastVerifiedBlockNumber, nil
```

Reply:
Good point!